micrometer-core/src/main/java/io/micrometer/core/instrument/binder/httpcomponents/PoolingHttpClientConnectionManagerMetricsBinder.java

```diff
@@ -0,0 +1,204 @@
+/**
+ * Copyright 2019 Pivotal Software, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.httpcomponents;
+
+import io.micrometer.core.instrument.*;
+import io.micrometer.core.instrument.binder.MeterBinder;
+import io.micrometer.core.lang.NonNull;
+import org.apache.http.conn.HttpClientConnectionManager;
+import org.apache.http.conn.routing.HttpRoute;
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
+import org.apache.http.pool.PoolStats;
+
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * Collects metrics from a {@link PoolingHttpClientConnectionManager}.
+ *
+ * It monitors the overall connection pool and can also be used to monitor
+ * connection pools per route.
+ *
+ * @author Benjamin Hubert (benjamin.hubert@willhaben.at)
+ * @since 1.2.0
+ */
+public class PoolingHttpClientConnectionManagerMetricsBinder implements MeterBinder {
+
+    private static final String NAME_HTTPCLIENT_CONNECTION_POOL_TOTAL_MAX             = "httpcomponents.httpclient.pool.total.max";
+    private static final String NAME_HTTPCLIENT_CONNECTION_POOL_TOTAL_AVAILABLE       = "httpcomponents.httpclient.pool.total.available";
+    private static final String NAME_HTTPCLIENT_CONNECTION_POOL_TOTAL_LEASED          = "httpcomponents.httpclient.pool.total.leased";
+    private static final String NAME_HTTPCLIENT_CONNECTION_POOL_TOTAL_PENDING         = "httpcomponents.httpclient.pool.total.pending";
+    private static final String NAME_HTTPCLIENT_CONNECTION_POOL_DEFAULT_MAX_PER_ROUTE = "httpcomponents.httpclient.pool.route.max.default";
+    private static final String NAME_HTTPCLIENT_CONNECTION_POOL_ROUTE_MAX             = "httpcomponents.httpclient.pool.route.max";
+    private static final String NAME_HTTPCLIENT_CONNECTION_POOL_ROUTE_AVAILABLE       = "httpcomponents.httpclient.pool.route.available";
+    private static final String NAME_HTTPCLIENT_CONNECTION_POOL_ROUTE_LEASED          = "httpcomponents.httpclient.pool.route.leased";
+    private static final String NAME_HTTPCLIENT_CONNECTION_POOL_ROUTE_PENDING         = "httpcomponents.httpclient.pool.route.pending";
```

Comment:
Wouldn't it be better to define metric name just as "httpcomponents.httpclient.pool" and let the rest be just tags? What allows users of this metric to define hierarchical metric names as they need/wish using micrometer's name mappers.
The tags would be:
total/route/max/available/leased/pending/max-default

Reply:
I'm not too familiar with the connection pools, however you want to treat tags a parts of a whole. For example is available+leased+pending = total then we would use those tags and eliminate total as a tag (or keep total as a separate name). 

Using that logic is how you would decide if 'max' would belong in the same metric name as a tag. I suspect max is a high water mark, or max available so mixing it in as a tag probably wouldn't be a good idea.

Reply:
The values I named `total` are the values you get with `getTotalStats`. `total` is **not** the sum of `available` + `leased` + `pending`, but as far I understand, its values should be the sum of the corresponding `route` stats. In other words:

 * sum of all `route.max` = `total.max`
 * sum of all `route.leased` = `total.leased`
 * ...

So, it might be possible to remove `total` stats and calculate its values using all `route` stats, but as noted in line 176, you should not always enable route stats. Thus, you don't always have stats about all routes.

So, I'd personally keep the total stats.

----

Yes, `max` is the configured maximum, so a high water mark, which I would keep as separate metric name. Same for `route.max.default`.

----

Regarding tags vs. metrics for `available`, `leased` and `pending`: These numbers come from [PoolStats](https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/pool/PoolStats.html), which describes them as follows:

 * `available` ... Gets the number idle **persistent connections**.
 * `leased` ... Gets the number of **persistent connections** tracked by the connection manager currently being used to execute requests.
 * `pending` ... Gets the number of **connection requests** being blocked awaiting a free connection.

So, it might make sense to publish `available` and `leased` together in one metric with tags, because both are a count of `persistent connections`. If I get it right, `pending` does **not** count the is the number of `persistent connections`, so I think this should stay as a separate metric.

----

What do you think now? Should I change something?

Reply:
Thanks for the explanations of the various metrics. The only change I would recommend is to capture your explanations in the `description` of the various gauge builders, or via comments in the code.

I like how adding `available` and `leased` as tags could express they are part of the same pool. You could also get a similar message by unifying their metrics prefixes to include 'persistent'. I would favor tags for that case, but it wouldn't give me heartburn if you didn't:

```
httpcomponents.httpclient.pool.persistent.available
httpcomponents.httpclient.pool.persistent.leased
```